src/engine.rs

```diff
@@ -126,7 +139,13 @@ pub fn run(params: FunctionRunParams) -> Result<FunctionRunResult> {
     let module = Module::from_file(&engine, &function_path)
         .map_err(|e| anyhow!("Couldn't load the Function {:?}: {}", &function_path, e))?;
 
-    let input_stream = MemoryInputPipe::new(input.raw.clone());
+    let uses_msgpack = uses_msgpack_provider(&module);
+    let input_stream = if input.codec == Codec::Messagepack || uses_msgpack {
+        let msgpack_bytes = rmp_serde::to_vec(&input.json_value.as_ref().unwrap())?;
+        MemoryInputPipe::new(msgpack_bytes)
+    } else {
+        MemoryInputPipe::new(input.raw.clone())
+    };
```

Comment:
The `input` in `FunctionRunParams` should already contain the messagepack serialized bytes, therefore I don't think we should be serializing here again. I think that what we should do instead is try to unify how the relationship between the `input` `BytesContainer` and the `import_modules` function, to ensure that we only ever do the serialization/deserialization/byte handling in the `BytesContainer` struct.

Reply:
I think in general what we want here is to modify the input `BytesContainer` and set the codec to messagepack if the module uses messagepack. So the inferring should happen right here https://github.com/Shopify/function-runner/blob/main/src/main.rs#L138 -- when  with that it'll be enough to ensure that the input and output bytes both are serialized correctly, and should solve the issue that you're experiencing in https://github.com/Shopify/function-runner/pull/460#discussion_r2060660596

Reply:
Alternatively, if this is to early in the process, I think making `FunctionRunParams` take in a `&mut BytesContainer` could also work. Then, later down the line, we could mutate the bytes container with the right codec.